graphql_compiler/schema/schema_info.py

```diff
@@ -1,13 +1,192 @@
 # Copyright 2019-present Kensho Technologies, LLC.
 from collections import namedtuple
+from dataclasses import dataclass
+from functools import partial
+from typing import Dict, Optional
 
+from graphql.type import GraphQLSchema
 from graphql.type.definition import GraphQLInterfaceType, GraphQLObjectType
 import six
 import sqlalchemy
+from sqlalchemy.dialects.mssql import dialect as mssql_dialect
+from sqlalchemy.dialects.mysql import dialect as mysql_dialect
+from sqlalchemy.dialects.postgresql import dialect as postgresql_dialect
+from sqlalchemy.engine.interfaces import Dialect
 
 from . import is_vertex_field_name
 
 
+# Describes the intent to join two tables using the specified columns.
+#
+# The resulting join expression could be something like:
+# JOIN origin_table.from_column = destination_table.to_column
+#
+# The type of join (inner vs left, etc.) is not specified.
+# The tables are not specified.
+DirectJoinDescriptor = namedtuple(
+    "DirectJoinDescriptor",
+    (
+        "from_column",  # The column in the source table we intend to join on.
+        "to_column",  # The column in the destination table we intend to join on.
+    ),
+)
+
+
+@dataclass
+class GenericSchemaInfo:
+    """Class for storing generic schema info required for querying."""
+
+    schema: GraphQLSchema
+
+    # Optional dict of GraphQL interface or type -> GraphQL union.
+    # Used as a workaround for GraphQL's lack of support for
+    # inheritance across "types" (i.e. non-interfaces), as well as a
+    # workaround for Gremlin's total lack of inheritance-awareness.
+    # The key-value pairs in the dict specify that the "key" type
+    # is equivalent to the "value" type, i.e. that the GraphQL type or
+    # interface in the key is the most-derived common supertype
+    # of every GraphQL type in the "value" GraphQL union.
+    # Recursive expansion of type equivalence hints is not performed,
+    # and only type-level correctness of this argument is enforced.
+    # See README.md for more details on everything this parameter does.
+    # *****
+    # Be very careful with this option, as bad input here will
+    # lead to incorrect output queries being generated.
+    # *****
+    # TODO: make sure we are treating empty dict the same as None
+    type_equivalence_hints: Optional[Dict[str, str]]
+
+
+@dataclass
+class GenericSchemaInfoOnly:
```

Comment:
My impression was that this was going to be an abstract base class, but it doesn't seem to set `metaclass=ABCMeta`.

If making it an ABC, then I'd name it something like `BackendSpecificSchemaInfo` instead, since its current name is somewhat confusing.